pkg/controller/controller.go

```diff
@@ -0,0 +1,264 @@
+func (c *Controller) GetSuccessCount() int {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	return c.successCount
+}
+
+// GetFailureCount returns the number of failed controller runs
+func (c *Controller) GetFailureCount() int {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	return c.failureCount
+}
+
+// GetLastError returns the last error returned
+func (c *Controller) GetLastError() error {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	return c.lastError
+}
+
+// GetLastErrorTimestamp returns the last error returned
+func (c *Controller) GetLastErrorTimestamp() time.Time {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	return c.lastErrorStamp
+}
+
+func (c *Controller) runController() {
+	errorRetries := 1
+
+	for {
+		var (
+			err      error
+			interval = c.RunInterval
+		)
+
+		if c.DoFunc != nil {
+			err = c.DoFunc()
+		} else {
+			err = fmt.Errorf("DoFunc is nil")
+		}
+
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				fieldControllerName:    c.Name,
+				fieldUUID:              c.uuid,
+				fieldConsecutiveErrors: errorRetries,
+			}).WithError(err).Debug("Controller run failed")
+
+			c.mutex.Lock()
+			c.lastError = err
+			c.lastErrorStamp = time.Now()
+			c.failureCount++
+			c.mutex.Unlock()
+
+			if !c.NoErrorRetry {
+				if c.ErrorRetryBaseDuration != time.Duration(0) {
+					interval = time.Duration(errorRetries) * c.ErrorRetryBaseDuration
+				} else {
+					interval = time.Duration(errorRetries) * time.Second
+				}
+
+				errorRetries++
+			}
+		} else {
+			c.mutex.Lock()
+			c.successCount++
+			c.mutex.Unlock()
+
+			// reset error retries after successful attempt
+			errorRetries = 1
+
+			// If no run interval is specified, no further updates
+			// are required and we can shut down the controller
+			if c.RunInterval == time.Duration(0) {
+				goto shutdown
+			}
+		}
+
+		select {
+		case <-c.stop:
+			goto shutdown
+
+		case <-time.After(interval):
+		}
+
+	}
+
+shutdown:
+	log.WithFields(logrus.Fields{
+		fieldControllerName: c.Name,
+		fieldUUID:           c.uuid,
+	}).Debug("Shutting down controller")
+}
+
+func (c *Controller) stopController() {
+	close(c.stop)
+}
+
+type controllerMap map[string]*Controller
+
+// Manager is a list of controllers
+type Manager struct {
+	controllers controllerMap
+	mutex       lock.RWMutex
+}
+
+// UpdateController installs or updates a controller in the manager. A
+// controller is identified by its name. If a controller with the name already
+// exist, the controller will be shut down and replaced with the provided
+// controller. Updating a controller will cause the DoFunc to be run
+// immediately regardless of any previous conditions. It will also cause any
+// statistics to be reset.
+func (m *Manager) UpdateController(ctrl *Controller) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if m.controllers == nil {
+		m.controllers = controllerMap{}
+	}
+
+	if oldCtrl, ok := m.controllers[ctrl.Name]; ok {
+		oldCtrl.stopController()
```

Comment:
This will race with re-assigning ctrl.stop and ctrl.uuid below. Both are accessed via the struct pointer in the runController function. If you store a reference to the channel and uuid in the scope of the function, that would sidestep the issue. Alternatively, disallow updating a controller (as a wrapper around a goroutine I'm not sure of the use case where updating it in place is needed).

Reply:
Yeah, this is why it's still wip. I've had the Controller argument be a copy initially but the unit tests benefit from checking the internal structure values. I'll split the Controller into a parameters structure so there is no problem.

Reply:
oops, it is wip! sorry!